networkx/algorithms/community/demon.py

```diff
@@ -0,0 +1,296 @@
+    for n in nx.nodes(ego_minus_ego):
+        if len(nx.neighbors(ego_minus_ego, n)) == 0:
+            ego_minus_ego.node[n]['communities'] = [n]
+
+        c_n = ego_minus_ego.node[n]['communities']
+
+        for c in c_n:
+            if c in community_to_nodes:
+                community_to_nodes[c].append(n)
+            else:
+                community_to_nodes[c] = [n, ego]
+
+    return community_to_nodes
+
+
+def _merge_communities(communities, actual_community, epsilon):
+    """
+    Builds communities using the incomplete communities detected by the label
+    propagation algorithm in the EgoMinusEgo graphs.
+
+    Parameters
+    ----------
+    communities: dict
+        Dictionary of communities
+
+    actual_community: set
+        A community
+
+    epsilon: real
+        The tolerance required in order to merge communities.
+
+    Returns
+    -------
+    communities : list
+        Merged communities
+    """
+
+    # if the community is already present return
+    if actual_community in communities:
+        return communities
+
+    else:
+        # search a community to merge with
+        inserted = False
+
+        for test_community in communities:
+            union = _generalized_inclusion(actual_community, test_community, epsilon)
+            # community to merge with found
+            if union:
+                communities.pop(communities.index(test_community))
+                communities.append(union)
+                inserted = True
+                break
+
+        # not merged: insert the original community
+        if not inserted:
+            communities.append(actual_community)
+
+    return communities
+
+
+def _generalized_inclusion(c1, c2, epsilon):
+    """
+    As explained in [1]_:
+    "Two  communities C and I are  merged  if  and  only  if  at most the
+    epsilon % of the smaller one is not included in the bigger one;  in  this
+    case, C and I are removed from the community list and their union is added
+    to the result set. The epsilon factor is introduced to vary the percentage
+    of common elements provided from each couple of communities: epsilon = 0
+    ensure that two communities are merged only if one of them is a proper
+    subset of the other, on the other hand with a value of epsilon = 1 even
+    communities that do not share a single node are merged together"
+
+    Parameters
+    ----------
+    c1 : set
+        Community 1.
+
+    c2 : set
+        Community 2.
+
+    epsilon: real
+        The tolerance required in order to merge communities
+
+    Returns
+    -------
+    union : set
+        Union of both communities if the similarity is greater than epsilon
+
+    References
+    ----------
+    .. [1] Michele Coscia, Giulio Rossetti, Fosca Giannotti, Dino Pedreschi:
+           DEMON: a local-first discovery method for overlapping communities.
+           KDD 2012:615-623.
+    """
+    intersection = c2 & c1
+    smaller_set = min(len(c1), len(c2))
```

Comment:
This variable name is misleading; a better name would be `smaller_set_size` or something.